core/src/com/unciv/logic/BarbarianManager.kt

```diff
@@ -0,0 +1,265 @@
+package com.unciv.logic
+
+import com.badlogic.gdx.math.Vector2
+import com.unciv.Constants
+import com.unciv.logic.civilization.NotificationIcon
+import com.unciv.logic.map.TileInfo
+import com.unciv.logic.map.TileMap
+import com.unciv.models.metadata.GameSpeed
+import java.util.*
+import kotlin.math.min
+import kotlin.math.pow
+
+class BarbarianManager {
+    val camps = ArrayList<Encampment>()
+
+    @Transient
+    lateinit var gameInfo: GameInfo
+
+    @Transient
+    lateinit var tileMap: TileMap
+
+    fun clone(): BarbarianManager {
+        val toReturn = BarbarianManager()
+        toReturn.camps.addAll(camps.map { it.clone() } )
+        return toReturn
+    }
+
+    fun setTransients(gameInfo: GameInfo) {
+        this.gameInfo = gameInfo
+        this.tileMap = gameInfo.tileMap
+
+        // Add any preexisting camps as Encampment objects
+        for (tile in tileMap.values) {
+            if (tile.improvement == Constants.barbarianEncampment
+                && camps.none { it.position == tile.position }) {
+                val newCamp = Encampment()
+                newCamp.position = tile.position
+                newCamp.gameInfo = gameInfo
+                camps.add(newCamp)
+            }
+        }
+
+        for (camp in camps)
+            camp.gameInfo = gameInfo
+    }
+
+    fun updateEncampments() {
+        // Check if camps were destroyed
+        for (camp in camps.toList()) {
+            if (tileMap[camp.position].improvement != Constants.barbarianEncampment) {
+                camps.remove(camp)
+            }
+        }
+
+        // Possibly place a new encampment
+        placeBarbarianEncampment()
+
+        // Update all existing camps
+        for (camp in camps) {
+            camp.update()
+        }
+    }
+
+    /** Called when an encampment was attacked, will speed up time to next spawn */
+    fun campAttacked(position: Vector2) {
+        camps.firstOrNull { it.position == position }?.wasAttacked()
+    }
+
+    fun placeBarbarianEncampment() {
+        // Before we do the expensive stuff, do a roll to see if we will place a camp at all
+        if (gameInfo.turns > 1 && Random().nextBoolean())
+            return
+
+        // Barbarians will only spawn in places that no one can see
+        val allViewableTiles = gameInfo.civilizations.filterNot { it.isBarbarian() || it.isSpectator() }
+            .flatMap { it.viewableTiles }.toHashSet()
+        val fogTiles = tileMap.values.filter { it.isLand && it !in allViewableTiles }
+
+        val fogTilesPerCamp = (tileMap.values.size.toFloat().pow(0.4f)).toInt() // Approximately
+
+        // Check if we have more room
+        var campsToAdd = (fogTiles.size / fogTilesPerCamp) - camps.size
+
+        // First turn of the game add 1/3 of all possible camps
+        if (gameInfo.turns == 1) {
+            campsToAdd /= 3
```

Comment:
shouldn't this round up so an original campsToAdd == 2 wouldn't end up doing nothing on turn 1?